### PR TITLE
xdpdump: make xdpdump aware of XDP_MODE_HW

### PR DESCRIPTION
**Pull request by chaudron**
This fixes issue #84

Signed-off-by: Eelco Chaudron <echaudro@redhat.com>

### CODE REVIEW

File: xdp-dump/xdpdump.c [hunk 1]
```diff
@@ -1603,14 +1603,23 @@ static bool capture_on_interface(struct dumpopt *cfg)
 	struct capture_programs      tgt_progs = {};
 
 	mp = xdp_multiprog__get_from_ifindex(cfg->iface.ifindex);
-	if (IS_ERR_OR_NULL(mp)) {
+	if (IS_ERR_OR_NULL(mp) ||
+	    (!IS_ERR_OR_NULL(mp) && xdp_multiprog__main_prog(mp) == NULL)) {
```
Comment (tohojo) [inline xdp-dump/xdpdump.c:1607]: Boolean logic short-circuiting makes that second `!IS_ERR_OR_NULL()` redundant...

File: xdp-dump/xdpdump.c [hunk 2]
```diff
@@ -1603,14 +1603,23 @@ static bool capture_on_interface(struct dumpopt *cfg)
...
+			xdp_multiprog__close(mp);
+			mp = NULL;
+		}
+
 		if (!cfg->load_xdp) {
-			pr_warn("WARNING: Specified interface does not have an XDP "
-				"program loaded, capturing\n         in legacy mode!\n");
+			pr_warn("WARNING: Specified interface does not have an XDP program loaded%s,"
+				"\n         capturing in legacy mode!\n",
+				IS_ERR_OR_NULL(mp) ? "" : " in software");
```
Comment (tohojo) [inline xdp-dump/xdpdump.c:1617]: You just closed `mp` and set it to `NULL` above, so this `IS_ERR_OR_NULL()` will always be true...

File: xdp-dump/xdpdump.c [hunk 3]
```diff
@@ -1603,14 +1603,23 @@ static bool capture_on_interface(struct dumpopt *cfg)
...
 			return capture_on_legacy_interface(cfg);
 		}
-		pr_warn("WARNING: Specified interface does not have an XDP program loaded!\n"
-			"         Will load a capture only XDP program!\n");
+		pr_warn("WARNING: Specified interface does not have an XDP program loaded%s!\n"
+			"         Will load a capture only XDP program!\n",
+			IS_ERR_OR_NULL(mp) ? "" : " in software");
```
Comment (tohojo) [inline xdp-dump/xdpdump.c:1622]: ...as will this one...

**Review by tohojo — CHANGES_REQUESTED** [review]
Some odd logic to fix... :)

File: xdp-dump/xdpdump.c (hunk 2 above)
Comment (chaudron) [reply xdp-dump/xdpdump.c:1617]: Messed this up, when I was adding the mp = NULL to fix another error corner case I ran into. Will fix it differently ;)

File: xdp-dump/xdpdump.c (hunk 1 above)
Comment (chaudron) [reply xdp-dump/xdpdump.c]: You are right, I normally rely on the compiler to do this. But in this case, it's obvious, will change.

**Comment by chaudron** [comment]
@tohojo new version is available...

**Comment by tohojo** [comment]
Yeah, that's better, thanks :)

**Pull request merged**